torch_np/_detail/_reductions.py

```diff
@@ -0,0 +1,164 @@
+    # XXX: this all should probably be generalized to a sum(a != 0, dtype=bool)
+    try:
+        return a.count_nonzero(axis)
+    except RuntimeError:
+        raise ValueError
+    return tensor
+
+
+def argmax(tensor, axis=None, out=None, *, keepdims=NoValue):
+    axis = _util.allow_only_single_axis(axis)
+    tensor = torch.argmax(tensor, axis)
+    return tensor
+
+def argmin(tensor, axis=None, out=None, *, keepdims=NoValue):
+    axis = _util.allow_only_single_axis(axis)
+    tensor = torch.argmin(tensor, axis)
+    return tensor
+
+
+def any(tensor, axis=None, *, where=NoValue):
+    if where is not None:
+        raise NotImplementedError
+
+    axis = _util.allow_only_single_axis(axis)
+
+    if axis is None:
+        result = tensor.any()
+    else:
+        result = tensor.any(axis)
+    return result
+
+
+def all(tensor, axis=None, *, where=NoValue):
+    if where is not None:
+        raise NotImplementedError
+
+    axis = _util.allow_only_single_axis(axis)
+
+    if axis is None:
+        result = tensor.all()
+    else:
+        result = tensor.all(axis)
+    return result
+
+
+def max(tensor, axis=None, initial=NoValue, where=NoValue):
+    if initial is not None or where is not None:
+        raise NotImplementedError
+
+    result = tensor.amax(axis)
+    return result
+
+
+def min(tensor, axis=None, initial=NoValue, where=NoValue):
+    if initial is not None or where is not None:
+        raise NotImplementedError
+
+    result = tensor.amin(axis)
+    return result
+
+
+def sum(tensor, axis=None, dtype=None, initial=NoValue, where=NoValue):
+    if initial is not None or where is not None:
+        raise NotImplementedError
+
+    assert dtype is None or isinstance(dtype, torch.dtype)
+
+    if dtype == torch.bool:
+        dtype = _scalar_types.default_int_type.dtype
+
+    if axis is None:
+        result = tensor.sum(dtype=dtype)
+    else:
+        result = tensor.sum(dtype=dtype, dim=axis)
+
+    return result
+
+
+def prod(tensor, axis=None, dtype=None, initial=NoValue, where=NoValue):
+    if initial is not None or where is not None:
+        raise NotImplementedError
+
+    axis = _util.allow_only_single_axis(axis)
+
+    if dtype == torch.bool:
+        dtype = _scalar_types.default_int_type.dtype
+
+    if axis is None:
+        result = tensor.prod(dtype=dtype)
+    else:
+        result = tensor.prod(dtype=dtype, dim=axis)
+
+    return result
+
+
+def mean(tensor, axis=None, dtype=None, *, where=NoValue):
+    if where is not None:
+        raise NotImplementedError
+
+    dtype = _atleast_float(dtype, tensor.dtype)
+
+    if axis is None:
+        result = tensor.mean(dtype=dtype)
+    else:
+        result = tensor.mean(dtype=dtype, dim=axis)
+
+    return result
+
+
+def std(tensor, axis=None, dtype=None, ddof=0, *, where=NoValue):
+    if where is not None:
+        raise NotImplementedError
+
+    dtype = _atleast_float(dtype, tensor.dtype)
+
+    tensor = tensor.to(dtype)
```

Comment:
Nit.
```suggestion
    if dtype != tensor.dtype:
        tensor = tensor.to(dtype)
```
This way we avoid having `to`'s scattered all throughout the trace of the program. In PrimTorch we have an aux function that does this called `_maybe_convert_to_dtype`.

Reply:
I was going to the other direction based on 

```
In [2]: t = torch.ones(4)

In [3]: t.to(t.dtype) is t
Out[3]: True
```

but OK, if it's preferable to have an explicit `if dtype != ...`, will add it everywhere.

Reply:
It'd be good to have a wrapper that does the cast iff the dtypes are the same. For eager, you are right, this is a noop. Now, when tracing through the program, these calls do show up, and would need to be cleaned up later on by the compiler. Guarding for this helps a bit the compiler / serialisation process in the long run.